File: lib/server/routes/converse.py
"""Conversational-engine routes: theme builder, onboarding, and Compose."""

from __future__ import annotations

import json
from typing import Any

from fastapi import APIRouter, Request, Response
from fastapi.responses import JSONResponse

from server import views

router = APIRouter()


@router.post("/api/theme-builder/converse", response_class=JSONResponse)
async def api_theme_builder_converse(request: Request) -> JSONResponse:
    """Run one turn of the conversational roster theme builder.

    Body: ``{ messages: [{role, content}] }``. Each call runs ONE assistant
    turn via the agent-engine dispatch, seeded with the theme-builder system
    prompt + the roster contract (role-slugs, role labels, current names) read
    server-side from ``roster_manifest.json``. The model asks a short vibe
    question, then proposes a full role-slug -> display-name mapping as a
    ``propose_theme`` action.

    Returns ``{ reply, action }`` where ``action`` is either ``null`` (a plain
    vibe-asking turn) or ``{tool: "propose_theme", args: {custom_names,
    custom_roles}}``. Nothing is saved here: the client pre-fills the custom
    theme editor with the proposal, the person confirms, and the client saves
    it via ``POST /api/roster-theme`` with ``theme: "custom"``. Degrades with a
    503 when no live engine is configured so the client falls back to the
    manual editor.
    """
    if not views._same_origin_post(request) or not views._authorized_mutation(request):
        return JSONResponse({"error": "forbidden"}, status_code=403)
    try:
        body = json.loads((await request.body()).decode("utf-8") or "{}")
    except (json.JSONDecodeError, UnicodeDecodeError):
        return JSONResponse({"error": "request body must be JSON"}, status_code=400)
    if not isinstance(body, dict):
        return JSONResponse({"error": "request body must be a JSON object"}, status_code=400)
    return views._run_theme_builder_converse(request, body)


@router.post("/api/onboarding/converse", response_class=JSONResponse)
async def api_onboarding_converse(request: Request) -> JSONResponse:
    """Run one turn of the conversational Ask-driven onboarding guide.

    Body: ``{ messages: [{role, content}] }``. Each call runs ONE assistant
    turn via the agent-engine dispatch, seeded with the onboarding system
    prompt. Alfred asks a short setup question, then REQUESTS a structured
    action (check the engines, connect GitHub, pick repos, name the team, set
    a schedule, finish) that the DESKTOP CLIENT executes under the same token
    gate the stepped flow uses. The model never writes config or a token: it
    only proposes the next step.

    Returns ``{ reply, action, done }`` where ``action`` is either ``null`` (a
    plain question turn) or ``{tool, args}`` for one scoped onboarding action.
    Nothing is executed here: the client runs the SAME setup handler the
    stepped OnboardingView already drives, so the two paths cannot drift.
    Degrades with a 503 when no live engine is configured so the client falls
    back to the stepped flow.
    """
    if not views._same_origin_post(request) or not views._authorized_mutation(request):
        return JSONResponse({"error": "forbidden"}, status_code=403)
    try:
        body = json.loads((await request.body()).decode("utf-8") or "{}")
    except (json.JSONDecodeError, UnicodeDecodeError):
        return JSONResponse({"error": "request body must be JSON"}, status_code=400)
    if not isinstance(body, dict):
        return JSONResponse({"error": "request body must be a JSON object"}, status_code=400)
    return views._run_onboarding_converse(request, body)


@router.post("/api/compose/converse", response_class=JSONResponse)
async def api_compose_converse(request: Request) -> JSONResponse:
    """Run one turn of the conversational, repo-grounded spec-builder.

    Body: ``{ draft_id?, context_repos?: [owner/repo], repos?: [owner/repo], plain?: bool,
    messages: [{role, content}] }``. Each call runs ONE assistant turn via
    the agent-engine dispatch, seeded with the spec-interrogator system
    prompt + repo grounding + code map. ``plain`` toggles jargon-free
    coaching for this turn (it wins over the ALFRED_INTAKE_PROFILE env
    default); the structured draft and readiness are unchanged either way.
    Persists the accumulating spec and conversation as a compose planning
    draft so it shows in Plans and threads into the RequestThread.

    Returns ``{ reply, draft, readiness:{score, ready, missing[]}, done }``.
    Degrades with a 503 when no live engine is configured (the off-Tauri
    browser preview never calls this; it stays on the one-shot rubric form).
    """
    if not views._same_origin_post(request) or not views._authorized_mutation(request):
        return JSONResponse({"error": "forbidden"}, status_code=403)
    try:
        body = json.loads((await request.body()).decode("utf-8") or "{}")
    except (json.JSONDecodeError, UnicodeDecodeError):
        return JSONResponse({"error": "request body must be JSON"}, status_code=400)
    if not isinstance(body, dict):
        return JSONResponse({"error": "request body must be a JSON object"}, status_code=400)
    return views._run_compose_converse(request, body)


@router.post("/api/compose/converse/stream")
async def api_compose_converse_stream(request: Request) -> Any:
    """Token-stream one converse turn so chat renders as the model writes (#36).

    Same body + auth + persistence as ``/api/compose/converse`` (a
    token-gated mutation). Because this is a POST it cannot ride
    ``EventSource`` (which is GET-only and cannot send ``X-Alfred-Token``);
    the native client consumes it via ``fetch()`` + a streamed
    ``ReadableStream``, which carries the token header. The response is an
    SSE byte stream:

    * ``open``   once, when the turn starts.
    * ``token``  each new assistant text fragment teed to the transcript
      (BEST EFFORT progress; a model that does not tee interim text simply
      emits none and the reply lands whole on ``result``).
    * ``result`` once, the full reconciled ``ConverseResponse`` (also
      persisted as a compose draft, exactly like the non-streaming route).
    * ``error``  when no live engine is configured or the turn failed, so
      the client falls back to the non-streaming converse / one-shot form.

    Auth is checked BEFORE the stream opens so a forbidden caller gets a
    clean 403 JSON, never a half-open stream.

    Unlike the buffered mutations, the packaged Tauri webview reaches this
    route cross-origin (its bundle loads from ``tauri://localhost``, not the
    server's Host), and it must to stream an incremental body the buffered
    Tauri JSON bridge cannot carry. So instead of strict same-origin we
    require (a) an allowed webview/localhost Origin AND (b) the per-launch
    token via ``_authorized_mutation`` (constant-time compare). The token is
    the real CSRF defense: a drive-by page cannot read the operator's
    ``0600`` token file, so a bare cross-origin POST without it is rejected.
    """
    cors = views._streaming_cors_headers(request)
    if (
        views._streaming_origin_allowed(request) is None and not views._same_origin_post(request)
    ) or not views._authorized_mutation(request):
        return JSONResponse({"error": "forbidden"}, status_code=403, headers=cors)
    try:
        body = json.loads((await request.body()).decode("utf-8") or "{}")
    except (json.JSONDecodeError, UnicodeDecodeError):
        return JSONResponse({"error": "request body must be JSON"}, status_code=400, headers=cors)
    if not isinstance(body, dict):
        return JSONResponse(
            {"error": "request body must be a JSON object"},
            status_code=400,
            headers=cors,
        )
    return views._stream_compose_converse(request, body)


@router.options("/api/compose/converse/stream")
async def api_compose_converse_stream_preflight(request: Request) -> Response:
    """CORS preflight for the cross-origin converse stream (#36).

    The packaged webview's token-bearing POST is a non-simple request, so
    the browser sends an ``OPTIONS`` preflight first. Answer it for allowed
    webview/localhost origins; this carries no body and runs no turn, so it
    is not token-gated (the actual POST still is).
    """
    return Response(status_code=204, headers=views._streaming_cors_headers(request))

File: clients/desktop/src/api/converse.ts
import { invoke } from "@tauri-apps/api/core";

import type {
  ComposeDraftRequest,
  ComposeDraftResponse,
  ConversationControlRequest,
  ConversationControlResponse,
  ConverseRequest,
  ConverseResponse,
} from "../types";
import {
  ApiError,
  hostedBrowserToken,
  humanizeFetchError,
  humanizeTransportError,
  isHostedBrowser,
  isTauri,
  readSseStream,
  serverErrorMessage,
  streamingUrl,
  writeAlfredJson,
} from "./client";

export async function composeDraft(
  baseUrl: string,
  request: ComposeDraftRequest,
): Promise<ComposeDraftResponse> {
  return writeAlfredJson(baseUrl, "/api/plans/draft", request);
}

export async function conversationControl(
  baseUrl: string,
  request: ConversationControlRequest,
  signal?: AbortSignal,
): Promise<ConversationControlResponse> {
  return writeAlfredJson(baseUrl, "/api/conversation/control", request, signal);
}

// One turn of the conversational, repo-grounded spec-builder. The server runs a
// single live interrogator turn and returns the reply + accumulating spec +
// readiness. When no live engine is configured the server returns a 503 with
// `error: "live_session_unavailable"`; the caller catches that (via
// isLiveSessionUnavailable) and degrades to the one-shot `composeDraft` form.
export async function composeConverse(
  baseUrl: string,
  request: ConverseRequest,
  signal?: AbortSignal,
): Promise<ConverseResponse> {
  return writeAlfredJson(baseUrl, "/api/compose/converse", request, signal);
}

// Token-stream one converse turn (#36). EventSource is GET-only and cannot send
// `X-Alfred-Token`, so this uses the webview's `fetch()` with a streamed
// ReadableStream body, which CAN carry the header. `onToken` fires with each
// assistant text fragment as it arrives; the returned promise resolves to the
// final reconciled ConverseResponse. On any streaming failure it REJECTS, so
// the caller can fall back to the non-streaming `composeConverse`. Native only
// (the browser preview stays on the one-shot form), but written transport-
// agnostically so it also works through the dev proxy.
export async function streamComposeConverse(
  baseUrl: string,
  request: ConverseRequest,
  onToken: (text: string) => void,
  signal?: AbortSignal,
): Promise<ConverseResponse> {
  const url = streamingUrl(baseUrl, "/api/compose/converse/stream");
  const headers: Record<string, string> = { "content-type": "application/json" };
  // Native build: attach the per-launch token the Rust side normally injects.
  // The dev proxy path is same-origin and the route still requires the token,
  // so attach it whenever we can read it; a missing token surfaces as the
  // server's 403, which the caller treats as a fallback trigger.
  if (isTauri()) {
    try {
      const token = await invoke<string>("alfred_server_token");
      if (token) {
        headers["X-Alfred-Token"] = token;
      }
    } catch {
      // No token: let the server reject so the caller falls back cleanly.
    }
  } else if (isHostedBrowser()) {
    // Browser served by `alfred serve`: the same-origin POST carries the
    // injected token. The stream route accepts same-origin + token, so this
    // works without the native bridge; a missing token surfaces as the 403 the
    // caller treats as a fallback trigger.
    const token = hostedBrowserToken();
    if (token) {
      headers["X-Alfred-Token"] = token;
    }
  }

  let response: Response;
  try {
    response = await fetch(url, {
      method: "POST",
      headers,
      body: JSON.stringify(request),
      signal,
    });
  } catch (err) {
    throw new ApiError(humanizeTransportError(err), err instanceof Error ? err.message : String(err));
  }
  if (!response.ok) {
    const text = await response.text().catch(() => "");
    const raw = `alfred serve returned ${response.status}${text ? `: ${text}` : ""}`;
    throw new ApiError(humanizeFetchError(response.status, serverErrorMessage(text)), raw);
  }

  let result: ConverseResponse | null = null;
  let streamError: string | null = null;
  for await (const frame of readSseStream(response, signal)) {
    if (frame.event === "token") {
      const text = (frame.data as { text?: string })?.text;
      if (typeof text === "string" && text) {
        onToken(text);
      }
    } else if (frame.event === "result") {
      result = frame.data as ConverseResponse;
    } else if (frame.event === "error") {
      streamError = (frame.data as { detail?: string })?.detail ?? "stream error";
    }
  }
  if (result) {
    return result;
  }
  // No result event: surface the server's degrade signal so the caller can fall
  // back to non-streaming converse (and then, if needed, the one-shot form).
  const detail = streamError ?? "live_session_unavailable";
  throw new ApiError("The conversational engine did not return a usable turn.", detail);
}
